android/BOINC/app/src/main/java/edu/berkeley/boinc/rpc/RpcClient.java

```diff
@@ -219,12 +263,20 @@ public synchronized void close() {
             if (Logging.WARNING) Log.w(Logging.TAG, "output close failure", e);
         }
         try {
-            mSocket.close();
+            if(mIsRemote){
```

Comment:
This is wrong. When called from `open()` function, (e.g. `mIsRemote == true`), here you'll try to close `mTcpSocket` but it could be wrong because at that moment could be active local connection and thus `mSocket` will still be open and connection is active that is wrong.  I'd suggest to remove `if(mIsRemote)` condition and try to close both sockets (yes, one of them will give error because it's not connected but it's fine and could be just logged as warning)

Reply:
When you switch from local socket to tcp socket, you must do it with runClient in monitor.
You cannot be just switch from Local to Tcp, it must be with ordered sequence:
Kill boinc, run it in the correct mode, connect it with the correct mode. All things are made from monitor.  
Decision is made in runClient in monitor and not from RpcClient. When you Kill boinc and run it again with runClient in monitor, it will update the remote by connect to the correct function of open() and update the mIsRemote accordingly.

Reply:
I remote the mIsRemote

Reply:
Idiot-proof design, you remember? ;)

Reply:
I don't see as Idiot-proof design, I see that we can determinate the system state in different way. so I remove the mIsRemote.